packages/driver/src/cy/commands/actions/attachFile.ts

```diff
@@ -0,0 +1,195 @@
+          onFail: options._log,
+          args: { file: JSON.stringify(file), index },
+        })
+      } else {
+        $errUtils.throwErrByPath('attachFile.invalid_single_file_reference', {
+          onFail: options._log,
+          args: { file: JSON.stringify(file) },
+        })
+      }
+    }
+
+    if (!_.isString(file.contents) && !Buffer.isBuffer(file.contents)) {
+      file.contents = JSON.stringify(file.contents)
+    }
+
+    return Promise.resolve(_.defaults({}, file, {
+      fileName: '',
+      lastModified: Date.now(),
+    }))
+  }
+}
+
+const createFileList = (files: Cypress.FileReferenceObject[]): FileList => {
+  const dataTransfer = new DataTransfer()
+
+  files.forEach(({ contents, fileName, lastModified, mimeType }) => {
+    const file = new File([contents], fileName || '', { lastModified, type: mimeType })
+
+    dataTransfer.items.add(file)
+  })
+
+  return dataTransfer.files
+}
+
+interface InternalAttachFileOptions extends Cypress.AttachFileOptions {
+  _log: any
+  $el: JQuery
+}
+
+export default (Commands, Cypress, cy, state, config) => {
+  Commands.addAll({ prevSubject: 'element' }, {
+    attachFile (subject: JQuery<any>, files: Cypress.FileReference | Cypress.FileReference[], options: Partial<InternalAttachFileOptions>): Promise<JQuery> {
+      options = _.defaults({}, options, {
+        action: 'input',
+        log: true,
+        $el: subject,
+      })
+
+      if (options.log) {
+        options._log = Cypress.log({
+          $el: options.$el,
+          timeout: options.timeout,
+          consoleProps () {
+            return {
+              'Target': $dom.getElements(options.$el),
+              Elements: options.$el?.length,
+            }
+          },
+        })
+
+        options._log.snapshot('before', { next: 'after' })
+      }
+
+      // TODO: Support for drag-n-drop to come at a later date
+      if (options.action !== 'input') {
+        $errUtils.throwErrByPath('attachFile.invalid_action', {
+          onFail: options._log,
+          args: { action: options.action },
+        })
+      }
+
+      if (subject.length > 1) {
+        $errUtils.throwErrByPath('attachFile.multiple_elements', {
+          onFail: options._log,
+          args: { num: subject.length },
+        })
+      }
+
+      let input = subject
+
+      if (input.is('label')) {
+        input = $dom.getInputFromLabel(input)
+      }
+
+      if (input.length < 1 || !$dom.isInputType(input, 'file')) {
+        const node = $dom.stringify(options.$el)
+
+        $errUtils.throwErrByPath('attachFile.not_file_input', {
+          onFail: options._log,
+          args: { node },
+        })
+      }
+
+      if (!options.force) {
+        cy.ensureNotDisabled(input, options._log)
+      }
+
+      // Make sure files is an array even if the user only passed in one
+      return Promise.all(([] as Cypress.FileReference[]).concat(files).map(parseFile(options))).then((filesArray) => {
```

Comment:
We could use async/await here and reduce some nesting. For new code, I think it's worth moving to async/await where possible.

Reply:
Agreed, I'm a big fan of async/await.